contracts/OffsetHelper.sol

```diff
@@ -296,102 +344,90 @@ contract OffsetHelper is OffsetHelperStorage {
-        if (_erc20Address == eligibleTokenAddresses["NCT"]) return true;
-        return false;
-    }
+        // auto redeem pool token for TCO2; will transfer automatically picked TCO2 to this contract
+        (tco2s, amounts) = PoolTokenImplementation.redeemAuto2(_amount);
 
-    /**
-     * @notice Return how much of the specified ERC20 token is required in
-     * order to swap for the desired amount of a Toucan pool token, for
-     * example, BCT or NCT.
-     *
-     * @param _fromToken The address of the ERC20 token used for the swap
-     * @param _toToken The address of the pool token to swap for,
-     * for example, NCT or BCT
-     * @param _toAmount The desired amount of pool token to receive
-     * @return amountsIn The amount of the ERC20 token required in order to
-     * swap for the specified amount of the pool token
-     */
-    function calculateNeededTokenAmount(
-        address _fromToken,
-        address _toToken,
-        uint256 _toAmount
-    ) public view onlySwappable(_fromToken) onlyRedeemable(_toToken) returns (uint256) {
-        (, uint256[] memory amounts) =
-            calculateExactOutSwap(_fromToken, _toToken, _toAmount);
-        return amounts[0];
+        // update balances
+        balances[msg.sender][_fromToken] -= _amount;
+        uint256 tco2sLen = tco2s.length;
+        for (uint256 index = 0; index < tco2sLen; index++) {
+            balances[msg.sender][tco2s[index]] += amounts[index];
+        }
+
+        emit Redeemed(msg.sender, _fromToken, tco2s, amounts);
     }
 
     /**
-     * @notice Calculates the expected amount of Toucan Pool token that can be
-     * acquired by swapping the provided amount of ERC20 token.
-     *
-     * @param _fromToken The address of the ERC20 token used for the swap
-     * @param _fromAmount The amount of ERC20 token to swap
-     * @param _toToken The address of the pool token to swap for,
-     * for example, NCT or BCT
-     * @return The expected amount of Pool token that can be acquired
+     * @notice Retire the specified TCO2 tokens.
+     * @param _tco2s The addresses of the TCO2s to retire
+     * @param _amounts The amounts to retire from each of the corresponding
+     * TCO2 addresses
      */
-    function calculateExpectedPoolTokenForToken(
-        address _fromToken,
-        uint256 _fromAmount,
-        address _toToken
-    ) public view onlySwappable(_fromToken) onlyRedeemable(_toToken) returns (uint256) {
-        (, uint256[] memory amounts) =
-            calculateExactInSwap(_fromToken, _fromAmount, _toToken);
-        return amounts[amounts.length - 1];
+    function autoRetire(
+        address[] memory _tco2s,
+        uint256[] memory _amounts
+    ) public {
+        uint256 tco2sLen = _tco2s.length;
+        require(tco2sLen != 0, "Array empty");
+
+        require(tco2sLen == _amounts.length, "Arrays unequal");
+
+        uint256 i = 0;
+        while (i < tco2sLen) {
```

Comment:
any reason to use while loops over for loops over the whole codefile?

Reply:
agree, was not my design. But replaced with for loop.